hilti/toolchain/src/compiler/optimizer.cc

```diff
@@ -977,6 +977,271 @@ struct ConstantFoldingVisitor : OptimizerVisitor {
     }
 };
 
+struct ConstantPropagationVisitor : OptimizerVisitor {
+    using OptimizerVisitor::OptimizerVisitor;
+    using OptimizerVisitor::operator();
+
+    struct ConstantValue {
+        Expression* expr;
+        // NAC
+        bool not_a_constant = false;
+
+        bool operator==(const ConstantValue& other) const {
+            return expr == other.expr && not_a_constant == other.not_a_constant;
+        }
+    };
+
+    using ConstantMap = std::map<Declaration*, ConstantValue>;
+    struct AnalysisResult {
+        detail::cfg::CFG cfg;
+        std::map<detail::cfg::GraphNode, ConstantMap> in;
+        std::map<detail::cfg::GraphNode, ConstantMap> out;
+    };
+
+    std::map<Node*, AnalysisResult> analysis_results;
+
+    void collect(Node* node) override {
+        _stage = Stage::COLLECT;
+        visitor::visit(*this, node);
+    }
+
+    bool prune_uses(Node* node) override {
+        _stage = Stage::PRUNE_USES;
+
+        clearModified();
+        visitor::visit(*this, node);
+
+        return isModified();
+    }
+
+    void transfer(const detail::cfg::GraphNode& n, ConstantMap& new_out) {
+        // Marks all children that are names as not a constant in the given map.
+        // This is used by function calls, since they have deeply nested names
+        // that should all be marked NAC.
+        struct NameNACer : hilti::visitor::PreOrder {
+            ConstantMap& constants;
+            NameNACer(ConstantMap& constants) : constants(constants) {}
+            void operator()(expression::Name* name) override {
+                if ( auto* decl = name->resolvedDeclaration() ) {
+                    constants[decl].not_a_constant = true;
+                }
+            }
+        };
+
+        struct TransferVisitor : hilti::visitor::PreOrder {
+            ConstantMap& constants;
+            NameNACer name_nac;
+
+            TransferVisitor(ConstantMap& constants) : constants(constants), name_nac(constants) {}
+
+            // Tries to evaluate an expression to a constant value given a map of known constants.
+            Expression* evaluate(Expression* expr) {
+                if ( expr->isA<expression::Ctor>() )
```

Comment:
Quick fly-by comment: not all ctors are constant. Sets for example are not. You probably want `expr->isConstant()` here, maybe in conjunction with the `Ctor` check, not sure.